src/Silverback.Integration/Messaging/Encryption/DecryptorConsumerBehavior.cs

```diff
@@ -38,9 +38,22 @@ public DecryptorConsumerBehavior(ISilverbackCryptoStreamFactory streamFactory)
             Check.NotNull(next, nameof(next));
 
             if (context.Envelope.Endpoint.Encryption != null &&
-                !context.Envelope.Headers.Contains(DefaultMessageHeaders.Decrypted) &&
+                !context.Envelope.Headers.Contains(DefaultMessageHeaders.Decrypted) && // TODO: Sergio when/where is it set?
                 context.Envelope.RawMessage != null)
             {
+                if (context.Envelope.Headers.Contains(DefaultMessageHeaders.EncryptionKeyId) &&
+                    (context.Envelope.Endpoint.Encryption is SymmetricEncryptionSettings settings &&
+                    settings.DecryptionKeyCallback != null))
+                {
+                    var keyIdentifier =
+                        context.Envelope.Headers.GetValue(DefaultMessageHeaders.EncryptionKeyId);
+
+                    if (!string.IsNullOrEmpty(keyIdentifier))
+                    {
+                        settings.Key = settings.DecryptionKeyCallback(keyIdentifier);
```

Comment:
Are you sure this approach is thread safe? ⚠️

Reply:
😲 ouch well here I need your help....

Reply:
The logic has been moved in `SymmetricDecryptStream.CreateCryptoTransform` method.